### DISCUSSION TITLE
Do concurrent client connections get created sequentially?

### DISCUSSION THREAD

**Discussion by hlecnt** [aio-libs/aiohttp#8170] [Q&A] [answered]
Hi all,

While investigating on a global performance issue on my aiohttp app (Python 3.10, aiohttp 3.9.3) I noticed that a significant amount of time is spent in the client session request() call when concurrent request are made. This was confirmed by enabling Tracing on some requests.

So I've developed a little docker app performing concurrent request to an Nginx server. It is available here: https://github.com/hlecnt/aiohttp-load-test-2/tree/master

By default it will generate 2 iterations of 500 concurrent requests to the Nginx server. 
A tcpdump is done during the test. 
The first iteration takes more time is than the second to complete. For me, this is expected because of the need to create connections with the upstream server.

But I observed that the first request is really sent only when all connections to the upstream server are established. And these connections are **established sequentially** (confirmed by the tcpdump). I have tested on a Google VM and a WSL intance with the same result.
Performing the same test by sending the request sequentially (in a for loop) instead of concurrently (gather()) gives better timings (TCPConnector reuse the same connection). This was unexpected but may be due to the test itself: nginx answers very quickly here.

I am wondering if this sequential creation of TCP connections is an expected behavior? 
_Personally I was expecting HTTP request sent before all the connections are setup. But this behavior may be due to the test configuration itself._

The second point is the total time spent to process the request, while it takes less than 1ms on the nginx side, it take about 500ms on my WSL to complete the 500 request. 
I know that I can't compare aiohttp with other server like nginx, but the total processing time seems to be 'too' high.

Any comment will be appreciated.

Best regards,

---

Please find attached an archive with 2 tcpdump (50 requests sent concurrently versus sequentially)
[dump-50-2iter.zip](https://github.com/aio-libs/aiohttp/files/14343469/dump-50-2iter.zip)

**Comment by bdraco** [comment]
When you do a gather there is a race to make connections to the remote host. 

The performance is frequently better when the server supports keep alive by awaiting requests in sequence (which has less task overhead for small requests) or using a semaphore to reduce the concurrency of tasks since it's more likely to see connection reuse.

**Reply by hlecnt** [reply]
@bdraco Thanks for your answer!

I understand that the sequential creation of the connection is due to the gather() and the nature of the test, this is fine.

About the second iteration, when connections are established and reused (confirmed by the tcpdump), it takes 400 ms instead of 500 ms (first iteration with cnx establishment) to complete the requests (gather used again here). 

This total time seems very high too here while connections are already established (keep alive), isn't ? But maybe this means that it takes almost 1ms to create a request in my env. As asyncio/python is using one thread at a time, it's normal to take 400ms to send all the 500 requests, correct ?

Regards

**Reply by bdraco** [reply]
That seems normal. I would run it all through cProfile to get a better idea where the time is being spent to see if there is any opportunity to optimize it.

**Reply by bdraco** [reply]
You might also try https://github.com/aio-libs/aiohttp/pull/8163 to see if it improves performance for your case

**Reply by bdraco** [reply]
write_bytes also runs in a task. So if you hit that path it's still going to have to suspend anyways.  Maybe we can avoid creating a task there as well if the payload is small and can be written immediately. That one is much harder to optimize away the task though because of the stream design.

**Reply by hlecnt** [reply]
I ran cProfile, but I can't spot something wrong. I post the `.prof` here if you have some time to quickly check it.

Thanks for your answers

---
[profile.zip](https://github.com/aio-libs/aiohttp/files/14345020/profile.zip)

**Reply by bdraco** [reply]
![create_task](https://github.com/aio-libs/aiohttp/assets/663432/99d68dc3-9d06-439a-98c6-40a41a9ed02f)
![fetch](https://github.com/aio-libs/aiohttp/assets/663432/11b36b8d-facf-497e-8d09-c9ca44dbbb09)

Looks to be a lot of asyncio overhead.

You might get a better result with python 3.12

There are a few patches to improve asyncio performance at https://github.com/home-assistant/docker-base/tree/master/python/3.12 that would likely help with your case

**Reply by hlecnt** [reply]
Many thanks for all this details, I will try python 3.12.

**Reply by hlecnt** [reply]
Hi, I made some other investigations. 

By instrumenting the code, I have noticed a lots of `RequestHandler._handle_request()` tasks during the test. 
The task creation is here:
https://github.com/aio-libs/aiohttp/blob/bf4edce5d6c967fa1d6a581b3aaab1bd2f5121cd/aiohttp/web_protocol.py#L527

My aiohttp reference is still a `3.4.3 + Python 3.6`. I have already spot a performance issue for my app since aiohttp 3.5+ (https://github.com/aio-libs/aiohttp/discussions/6008) and I was waiting for v3.9.3 to be released to move on.

When comparing `RequestHandler.start()` in aiohttp 3.4.3 with aiohttp 3.9.3, one difference is the use of `create_task()` instead of an `await` in order to copy context vars.

I had a look to https://stackoverflow.com/questions/55761652/what-is-the-overhead-of-an-asyncio-task and I wrote a small script in order to simulate the change from `await` to `create_task`
```
import asyncio


async def mytask():
    await asyncio.sleep(10)


async def handle_request():
    pass


async def start34():
    await handle_request()


async def processing34():
    t = asyncio.get_event_loop().create_task(start34())
    await t


async def start39():
    t = asyncio.get_event_loop().create_task(handle_request())
    await t


async def processing39():
    t = asyncio.get_event_loop().create_task(start39())
    await t


async def main():
    N = 1
    t0 = asyncio.get_event_loop().time()
    tasks = [asyncio.get_event_loop().create_task(handle_request()) for _ in range(N)]
    t1 = asyncio.get_event_loop().time()
    await asyncio.gather(*tasks)
    t2 = asyncio.get_event_loop().time()
    print(f"create_task total_time={t2-t0}, create_time={t1-t0}, gather_time={t2-t1}")

    t0 = asyncio.get_event_loop().time()
    coros = [handle_request() for _ in range(N)]
    t1 = asyncio.get_event_loop().time()
    await asyncio.gather(*coros)
    t2 = asyncio.get_event_loop().time()
    print(f"coros total_time={t2-t0}, create_time={t1-t0}, gather_time={t2-t1}")

    N = 100000
    t0 = asyncio.get_event_loop().time()
    tasks = [asyncio.get_event_loop().create_task(processing34()) for _ in range(N)]
    t1 = asyncio.get_event_loop().time()
    await asyncio.gather(*tasks)
    t2 = asyncio.get_event_loop().time()
    print(f"processing34 total_time={t2-t0}, create_time={t1-t0}, gather_time={t2-t1}")

    t0 = asyncio.get_event_loop().time()
    tasks = [asyncio.get_event_loop().create_task(processing39()) for _ in range(N)]
    t1 = asyncio.get_event_loop().time()
    await asyncio.gather(*tasks)
    t2 = asyncio.get_event_loop().time()
    print(f"processing39 total_time={t2-t0}, create_time={t1-t0}, gather_time={t2-t1}")


if __name__ == "__main__":
    asyncio.get_event_loop().run_until_complete(main())
```

* processing34 => tries to simulate aiohttp 3.4
* processing39 => tries to simulate aiohttp 3.9

Here are the results on my env (4 cores GCP VM):
* python 3.6:
```
processing34 total_time=4.05664858199998, create_time=1.2554828189995533, gather_time=2.8011657630004265
processing39 total_time=6.257052848000058, create_time=1.3079723239998202, gather_time=4.949080524000237
```
* python 3.10: 
```
processing34 total_time=4.59320455299985, create_time=1.2294919360001586, gather_time=3.363712616999692
processing39 total_time=7.01707209200049, create_time=1.726248814000428, gather_time=5.290823278000062
```
* python 3.11:
```
processing34 total_time=3.2123291730003984, create_time=1.0118171170006462, gather_time=2.200512055999752
processing39 total_time=4.942378323999947, create_time=1.1279427210001813, gather_time=3.8144356029997653
```
* python 3.12:
```
processing34 total_time=3.2715305440005977, create_time=0.945816308000758, gather_time=2.3257142359998397
processing39 total_time=4.513056855000286, create_time=1.0630218220003371, gather_time=3.450035032999949
```

Even if since Python 3.11 the `create_task()` overhead has been reduced it's still greater than the await.

If this is relevant, I guess the performance slow down will appear during "high" concurrent requests handling the aiohttp server part.

What do you think about this ?

Thanks

**Reply by bdraco** [reply]
`create_task` has to be scheduled on the loop so its always going to take more walk clock time (and some cpu time) unless the task is created with `eager_start=True` but thats py3.12 only.

**Reply by bdraco** [reply]
I got rid of another task creation in #8163 for dns hit, but I think getting rid of the one that was introduced in #3406 is difficult to do.

**Reply by bdraco** [reply]
If you can do py3.12 you can use https://docs.python.org/3/library/asyncio-task.html#eager-task-factory. 

You could also borrow this helper https://github.com/home-assistant/core/blob/f3594c543d2a923b57218c13a3f0b5bb9927588f/homeassistant/util/async_.py#L27

**Reply by hlecnt** [reply]
Thanks for your answers!
It seems that I'm on the right way. I'll get a try to Python 3.11 first, but I need some time to migrate the code properly.

**Reply by hlecnt** [reply, marked as answer]
Hi, 

Just a word before closing the discussion. Moving to Python 3.11 and some optim in my code have fixed performance issues. 
Thanks for your help @bdraco.

**Reply by bdraco** [reply]
https://github.com/aio-libs/aiohttp/pull/8661 will take care of the one I mentioned above when using python 3.12+
